Initial dump of my code

mcharsley opened the pull request:
Initial dump of my code. Supports
 brackets
 commas
-o
-mtime/ctime/atime
-print
-name/iname
-type (just file and directory at the moment)
-newer
-prune
-maxdepth/mindepth
-depth
-true/-false
And has a reasonable amount of unit/integration tests.

Quick summary of the files:
 - find/main.rs is an absolutely minimal wrapper that creates some dependencies and then calls into the library
 - find/mod.rs is the top-level library that handles argument parsing and walking down the filesystem tree
 - find/matchers/mod.rs defines a Matcher trait that takes a DirEntry and says whether it matches an arbitrary criteria or not, and has various associated helper functions 
 - find/matchers/logical.rs has some Matcher implementations for ANDing and ORing collections of sub-matchers
 - find/matchers/*.rs the other files in this submodule are fairly obvious, containing matchers that check name, size, age etc.
 - test_data/* has some static files that are used by the tests to reduce the need for creating and tearing down temporary files for all the tests -  I really wish Rust had a way to target a fake in-memory filesystem, like Java 7's FileSystem interface :-(

I'll leave that for you to review, and in the meantime, I'd better get on with perf and polishing my Google Resume...

antiagainst commented on Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+walkdir = { git = "https://github.com/mcharsley/walkdir", rev = "dffefcf8db97a331a0f81d120e8aa20c1b36251e" }
+tempdir = "0.3"
+regex = "0.2"
+num-traits = "0.1.37"
```
Why requiring a patch number for `num-traits` but not others?

raphlinus replied:
Also, given that we are building binaries, shouldn't we check in a `Cargo.lock` as well?

antiagainst replied:
+1. That's recommended by cargo.

mcharsley replied:
That was what num-traits documentation recommended using. As it turns out I don't need num-traits any more, so I'm deleting it.

mcharsley replied:
Will add Cargo.lock and stop ignoring it

antiagainst commented on src/find/main.rs:
```diff
@@ -0,0 +1,9 @@
+extern crate glob;
```
These source files should have copyright headers.

raphlinus replied:
Also, the copyright in LICENSE should be to either "Google Inc." or to "the findutils project maintainers".

mcharsley replied:
Copyright headers? Eurgh :-( Is there a standard template I should be copying from?

Rather more embarrassing copyright line in LICENSE now corrected.

antiagainst commented on src/find/mod.rs:
```diff
@@ -0,0 +1,476 @@
+use walkdir::WalkDir;
+use walkdir::WalkDirIterator;
+
+pub struct Config {
```
`#derive[Default]` maybe? It's meaningful to have a default config.

mcharsley replied:
I could, but I'm struggling to see the benefit. max_depth absolutely should not default to 0. I could change 
```
impl Config {
    fn new() -> Config {
```
to 
```
impl Default for Config {
    fn default() -> Config {
```

But from what I can see I'd then have to change
`let config = Config::new()`
to
`let config : Config = Default::default()`
Which is more to type and requires `Default` to be `use`d. Happy to be persuaded if you can show me how it would be useful (I'm still learning the language after all)

antiagainst replied:
Sorry, my bad. Not `#derive[Default]` but `impl Default`. You can have both and in `impl Default` just call `new`. So used to write `#derive[...]`. :(

The benefit is that you can enable clients of `Config`, e.g., a struct having a `Config` as field, to be able to use `#derive[Default]`. https://doc.rust-lang.org/core/default/trait.Default.html#derivable

Another usage could be possible, say if I'm writing my configuration in a json file and then using serde to load that json file into some data structure like:

```rust
struct MyGigantConfig {
    #[serde(default)]
    findutils_config: findutils::..::Config
    // Some other configs
}
```
Which means just use the default value for `Config` if the json file doesn't specify it.

mcharsley replied:
Fair enough. And I realised I can just use `let config = Config::default()` which is less painful than `let config : Config = Default::default()`

antiagainst commented on src/find/mod.rs:
```diff
@@ -0,0 +1,476 @@
+    config: Config,
+}
+
+/// Function to generate a ParsedInfoi from the strings supplied on the command-line.
```
s/`ParsedInfoi`/`ParsedInfo`/

mcharsley replied:
Done

antiagainst commented on src/find/matchers/logical_matchers.rs:
```diff
@@ -0,0 +1,477 @@
+        AndMatcherBuilder { submatchers: Vec::new() }
+    }
+
+    pub fn new_and_condition(&mut self, matcher: Box<Matcher>) {
```
I would suggest just to call this as `and`. It's already concise and clear.

mcharsley replied:
This was originally called "and" and OrMatcherBuilder's equivalent was originally called "or". But then ListMatcherBuilder didn't have an obvious matching name... I'd like to keep the 3 methods fairly consistent name-wise but am open to suggestions

antiagainst replied:
That's fair.

antiagainst commented on src/find/matchers/logical_matchers.rs:
```diff
@@ -0,0 +1,477 @@
+
+    pub fn new_or_condition(&mut self, arg: &str) -> Result<(), Box<Error>> {
+        if self.submatchers.last().unwrap().submatchers.is_empty() {
+            return Err(From::from(format!("invalid expression; you have used a binary operator \
```
I would suggest to define our own `Error` enum so that this can be an enumerant. It's more structured that way. :)

mcharsley replied:
Can I add that as a TODO for a later CL?

antiagainst replied:
Of course!

antiagainst commented on src/find/matchers/logical_matchers.rs:
```diff
@@ -0,0 +1,477 @@
+            let grandchild_and_matcher = &child_or_matcher.submatchers.last().unwrap();
+
+            if grandchild_and_matcher.submatchers.is_empty() {
+                return Err(From::from("invalid expression; you have used a binary operator '-a' \
```
And this can be another error enumerant. Something for the rest.

antiagainst commented on src/find/matchers/mod.rs:
```diff
@@ -0,0 +1,732 @@
+mod logical_matchers;
```
Maybe considering make some of these modules as public (also in other modules) so we can actually call the functionality as a library? But definitely can wait.

antiagainst commented on src/find/matchers/mod.rs:
```diff
@@ -0,0 +1,732 @@
+pub fn build_top_level_matcher(args: &[&str],
+                               config: &mut Config)
+                               -> Result<Box<Matcher>, Box<Error>> {
+    let (_, top_level_matcher) = try!(build_matcher_tree(args, config, 0, false));
```
Can use `?` for `try!`. But not a big thing. :)

mcharsley replied:
I prefer try! - my eye skips over ? and I miss the implied possible early return. But my eye is inexperienced at rust, so if ? is the standard idiom then I can switch... Let me know

antiagainst replied:
Both are fine, I think.

raphlinus replied:
My personal feeling is that ? is now standard and should be preferred. This is a point of style, though.

antiagainst commented on src/find/matchers/mod.rs:
```diff
@@ -0,0 +1,732 @@
+                    "-mtime" => time::FileTimeType::Modified,
+                    // This shouldn't be possible. We've already checked the value
+                    // is one of those three values.
+                    _ => panic!("Encountered unexpected value {}", args[i]),
```
`unreachable!()`?

mcharsley replied:
Oooh: nice!

antiagainst commented on src/find/matchers/size.rs:
```diff
@@ -0,0 +1,174 @@
+}
+
+impl Unit {
+    fn from_string(s: &str) -> Result<Unit, Box<Error>> {
```
Maybe `impl FromStr for Unit`?

mcharsley replied:
Once again: nice!

antiagainst commented on src/find/matchers/type_matcher.rs:
```diff
@@ -0,0 +1,75 @@
+}
+
+impl TypeMatcher {
+    pub fn new(type_string: &str) -> Result<TypeMatcher, Box<Error>> {
```
`impl FromStr for TypeMatcher`?

mcharsley replied:
I prefer the consistency of every Matcher having a new method (and a new_box), rather than ones that take a single string using FromStr and all the others using new.

There's a difference between "this is an object that can be represented as a string" and "this is an object that happens to take a single string as an argument".

antiagainst replied:
It's not selecting one between `impl FromStr` and `pub fn new()`. Can have both and actually call `from_str()` in `new()`.

From the [spec](https://doc.rust-lang.org/std/str/trait.FromStr.html) of `FromStr`:

> A trait to abstract the idea of creating a new instance of a type from a string.

IMHO, that's what we are doing here, it's not "an object that can be represented as a string", which sounds more like `impl Display` for me.

Although I don't have strong examples to show the importance of having `impl FromStr`. It's basically conventional stuff, IMHO. So, entirely up to you. :)

antiagainst replied:
It's not selecting one between `impl FromStr` and `pub fn new()`. Can have both and actually call `from_str()` in `new()`.

From the [spec](https://doc.rust-lang.org/std/str/trait.FromStr.html) of `FromStr`:

> A trait to abstract the idea of creating a new instance of a type from a string.

IMHO, that's what we are doing here, it's not "an object that can be represented as a string", which sounds more like `impl Display` for me.

Although I don't have strong examples to show the importance of having `impl FromStr`. It's basically conventional stuff, IMHO. So, entirely up to you. :)

raphlinus commented on src/find/matchers/size.rs:
```diff
@@ -0,0 +1,174 @@
+    Byte,
+    TwoByteWord,
+    Block,
+    KiloByte,
```
I could probably answer this myself from looking at the rest of the code, but should it be "Kilo" or "Kibi"?

mcharsley replied:
Done

raphlinus commented on src/find/matchers/size.rs:
```diff
@@ -0,0 +1,174 @@
+    if divisor == 1 {
+        return byte_size;
+    }
+    // We want to round up (e.g. 1 byte - 1024 bytes = 1k.
```
Another way to express this (which would avoid the need to special-case 0) is `(byte_size + divisor - 1) / divisor`. For better efficiency, might be best to use lg2(divisor), and shift right by that.

mcharsley replied:
Normally I'd expect the compiler to optimise the division into right shifts, but this is getting sufficiently removed from just dividing by a constant that I'd understand if it failed.

Adding the divisor could potentially cause an overflow error for a spectacularly large file and short-cutting the entire function for 0 seems reasonable enough. If you disagree about the cost of pipeline stalls caused by conditionals and the relative risk of encountering exabyte-scale files I can change it though,

raphlinus replied:
The overflow can be solved by using .saturating_add() rather than simply +. It's a good point, though, you want to be very conscious of overflow behavior.

Because the divisor is dynamic, I wouldn't expect the optimization.

I don't have a strong feeling about any of this, just offering alternatives.

raphlinus commented on src/find/matchers/size.rs:
```diff
@@ -0,0 +1,174 @@
+                "512-byte file should match size of 1 block");
+    }
+
+
```
too many blank lines?

antiagainst replied:
I just wonder whether rustfmt will do the work to remove this.

mcharsley replied:
My old nemesis: vertical whitespace :-(

Alas rustfmt doesn't help here (My eclipse workspace runs it at every save)

raphlinus commented on src/find/matchers/time.rs:
```diff
@@ -0,0 +1,325 @@
+
+use find::matchers::{ComparableValue, Matcher, MatcherIO};
+
+const SECS_PER_DAY: i64 = 60 * 60 * 24;
```
I'd write out "SECONDS"

mcharsley replied:
Done

raphlinus commented on src/find/matchers/printer.rs:
```diff
@@ -0,0 +1,50 @@
+use find::matchers::{Matcher, MatcherIO};
+
+/// This matcher just prints the name of the file to stdout.
+pub struct Printer {
```
This can be written `pub struct Printer` (no braces).

mcharsley replied:
done

raphlinus commented on src/find/matchers/prune.rs:
```diff
@@ -0,0 +1,49 @@
+use find::matchers::{Matcher, MatcherIO};
+
+/// This matcher checks the type of the file.
+pub struct PruneMatcher {
```
Same comment as PrintMatcher.

mcharsley replied:
done

raphlinus commented on src/find/matchers/time.rs:
```diff
@@ -0,0 +1,325 @@
+    fn matches(&self, file_info: &DirEntry, _: &mut MatcherIO) -> bool {
+        match self.matches_impl(file_info) {
+            Err(e) => {
+                writeln!(&mut stderr(),
```
Consider a `print_err!` macro for error printing; among other things, this would make it easier to abstract the actual error logging.

mcharsley replied:
Can I leave that as a TODO (as you say abstracting out error reporting will be useful)?

raphlinus replied:
Sure, np.

raphlinus commented on src/find/matchers/time.rs:
```diff
@@ -0,0 +1,325 @@
+            }
+        };
+        let age_in_seconds: i64 = age.as_secs() as i64 * if is_negative { -1 } else { 1 };
+        // rust division truncates towards zero (see
```
This logic isn't exactly truncation, it's "age in days" - 1 if the age is an exact multiple. Maybe that's ok, but comment doesn't quite match code.

mcharsley replied:
Have changed the comment. Hope it's more clear now.

raphlinus commented on src/find/mod.rs:
```diff
@@ -0,0 +1,476 @@
+                   config: &Config,
+                   deps: &'a Dependencies<'a>,
+                   matcher: &Box<matchers::Matcher>)
+                   -> Result<i32, Box<Error>> {
```
Any particular reason this is i32, as opposed to, say, isize?

mcharsley replied:
Force of habit. Will change to a u64 just to be safe.

raphlinus commented on src/find/mod.rs:
```diff
@@ -0,0 +1,476 @@
+
+If no path is supplied then the current working directory is used by default.
+
+Early alpha implementation. Currently the only expressions supported are
```
Is this comment stale? It seems like you have more matchers defined.

mcharsley replied:
It is indeed hopelessly out of date. Fixing now

raphlinus commented on src/find/matchers/logical_matchers.rs:
```diff
@@ -0,0 +1,477 @@
+/// side effects, the side effects occur in the same order as the sub-matchers
+/// were pushed into the collection.
+pub struct AndMatcher {
+    submatchers: Vec<Box<Matcher>>,
```
Vec<Matcher> should work ok here, not sure why you're adding an additional `Box` indirection.

raphlinus commented on src/find/matchers/logical_matchers.rs:
```diff
@@ -0,0 +1,477 @@
+
+
+/// A simple matcher that always matches.
+pub struct TrueMatcher {
```
Can also be unit-like struct.

mcharsley replied:
done

raphlinus commented on src/find/matchers/logical_matchers.rs:
```diff
@@ -0,0 +1,477 @@
+}
+
+/// A simple matcher that never matches.
+pub struct FalseMatcher {
```
Can also be unit-like struct.

mcharsley replied:
done

mcharsley commented:
Thanks all. As a beginner with git, what's the procedure here? Do I make
changes to my tree and create a new pull request? Or is there something
clever I can do with the existing one?

Thanks

Mark

On 1 Mar 2017 6:55 pm, "Raph Levien" <notifications@github.com> wrote:

> *@raphlinus* commented on this pull request.
>
> Overall looks good. Definitely an advanced use of a trait for dependencies
> for easier testing, which I think is a very good thing here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/findutils/pull/1#pullrequestreview-24559407>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALbIF86FcmqvDrwUayrYNFuTvzDyXUL3ks5rhb8ogaJpZM4MPdz4>
> .
>

raphlinus commented:
The usual flow on github is to do an additional commit with the new changes and push that to your branch (mcharsley:master). That'll automatically update the pull request. Then one of us clicks "merge pull request" in the web UI when we approve.

antiagainst commented:
I would typically prefer "squash & merge" since that hides rounds of iteration details (like fixing typo, addressing comments, significant or not) and make the history more clean. :)

raphlinus commented:
I actually prefer squash and merge (it's much closer to gerrit-style workflows), but almost none of the github-hosted projects I work on use it.

antiagainst commented:
@raphlinus: Haha, we did that for all the projects we maintain: https://github.com/google/shaderc and https://github.com/KhronosGroup/SPIRV-Tools. ;-P

mcharsley commented on .settings/com.github.rustdt.ide.core.prefs:
```diff
@@ -0,0 +1,2 @@
+build_targets=<?xml version\="1.0" encoding\="UTF-8" standalone\="no"?>\n<build_targets xmlns\="com.github.rustdt.ide.core">\n<target auto_enabled\="false" config\="build" n_enabled\="true" version2\="true"/>\n<target auto_enabled\="true" config\="check" n_enabled\="false" version2\="true"/>\n</build_targets>\n
```
That shouldn't have been commited. I'll remove it and git-ignore it.

mcharsley commented:
OK changes committed and pushed. Hopefully you can see them

Mark

On Wed, Mar 1, 2017 at 7:16 PM, Raph Levien <notifications@github.com>
wrote:

> The usual flow on github is to do an additional commit with the new
> changes and push that to your branch (mcharsley:master). That'll
> automatically update the pull request. Then one of us clicks "merge pull
> request" in the web UI when we approve.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/findutils/pull/1#issuecomment-283439882>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALbIF1vn4bs9Pn8cjkNVpBnmWIZAMb3lks5rhcP8gaJpZM4MPdz4>
> .
>

mcharsley commented:
I'm a big fan of dependency injection. Less of a fan of people being
"clever" with Spring and Guice though.

Mark

On Wed, Mar 1, 2017 at 6:55 PM, Raph Levien <notifications@github.com>
wrote:

> *@raphlinus* commented on this pull request.
>
> Overall looks good. Definitely an advanced use of a trait for dependencies
> for easier testing, which I think is a very good thing here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/findutils/pull/1#pullrequestreview-24559407>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALbIF86FcmqvDrwUayrYNFuTvzDyXUL3ks5rhb8ogaJpZM4MPdz4>
> .
>

raphlinus commented:
I have no concerns that would block merging this, but won't merge right now in case Mark wants to do minor touchups based on the conversation. Feel free to merge.

antiagainst commented:
Also +1 from me. :)

mcharsley commented:
I've changed the try! macros to ? and added copyright headers. I'm happy for your guys to merge at your convenience :-)

Next step is probably -exec but I've got to spend time preparing for my SETI -> SWE transition first :-(

mcharsley commented:
Friendly ping. Can someone press the button to do the merge? Or am OK to do it? In which case, should I use
 - Create a merge commit
 - Squash and merge
 - Rebase and merge
 - something else?

thanks

mark

antiagainst commented:
Hi @mcharsley, definitely feel free to merge it! I'd recommend "squash and merge".
